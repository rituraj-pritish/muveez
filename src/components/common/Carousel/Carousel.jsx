import React from 'react'
import PropTypes from 'prop-types'
import 'slick-carousel/slick/slick.css'
import 'slick-carousel/slick/slick-theme.css'
import Slider from 'react-slick'

import { ReactComponent as ChevronRight } from 'assets/icons/chevron-right.svg'
import { ReactComponent as ChevronLeft } from 'assets/icons/chevron-left.svg'
import Card from '../Card'
import { SliderWrapper, Wrapper } from './Carousel.styles'
import Text from '../ui/Text'
import Toggle from '../Toggle'
import FlexBox from '../ui/FlexBox'
import Icon from '../ui/Icon'
import Link from '../ui/Link'

const DATA_CHUNK = 20

const Carousel = ({ 
	title, 
	data, 
	toggleOptions, 
	onToggleChange, 
	containerRef, 
	viewAllLink,
	cardRender,
	isLoading,
	...rest
}) => {
	const settings = {
		lazyLoad: true,
		swipeToSlide: true,
		arrows: data ? true : false,
		rows: 1,
		infinite: true,
		speed: 500,
		slidesToShow: 5,
		slidesToScroll: 5,
		nextArrow: (
			<Icon
				onClick={() => console.log('yessss caleda;sdlfkj')}
				className='slick-arrow slick-next'
				color='white'
				size={30}
				data-testid='slick-next'
			>
				<ChevronRight/>
			</Icon>
		),
		prevArrow: (
			<Icon
				className='slick-arrow slick-prev'
				color='white'
				size={30}
				data-testid='slick-prev'
			>
				<ChevronLeft/>
			</Icon>
		),
		responsive: [
			{
				breakpoint: 1150,
				settings: {
					slidesToShow: 4,
					slidesToScroll: 4
				}
			},
			{
				breakpoint: 900,
				settings: {
					slidesToShow: 3,
					slidesToScroll: 3
				}
			},
			{
				breakpoint: 600,
				settings: {
					slidesToShow: 2,
					slidesToScroll: 2
				}
			}
		]
	}

	const render = () => {
		if(isLoading || !data) return new Array(5)
			.fill(0).map((_, idx) => <Card key={idx} />)

		if(Array.isArray(data) && data.length === 0) return (
			<div>
				<FlexBox
					alignItems='center'
					height='100%'
				>
					No data available
				</FlexBox>
			</div>
		)

		return data.slice(0, DATA_CHUNK).map((item) => cardRender
			? cardRender(item)
			:	(
				<Card
					key={item.id}
					{...item}
				/>)
		)
	}

	return (
		<Wrapper
			ref={containerRef}
			data-testid={`${title}-section`}
			{...rest}
		>
			<div/>
			<FlexBox
				alignItems='center'
				justifyContent='space-between'
				mb={3}
			>
				<div>
					<Text
						bold
						color='primary'
						size={4}
						mr={4}
					>
						{title}
					</Text>
					{toggleOptions && (
						<Toggle
							options={toggleOptions}
							onChange={onToggleChange}
						/>
					)}
				</div>
				{viewAllLink && (
					<Link
						size={1}
						to={viewAllLink}
					>
						View all
					</Link>
				)}
			</FlexBox>
			<SliderWrapper>
				<Slider {...settings}>
					{render()}
				</Slider>
			</SliderWrapper>
		</Wrapper>
	)
}

Carousel.propTypes = {
	title: PropTypes.string.isRequired,
	data: PropTypes.array,
	toggleOptions: PropTypes.array,
	onToggleChange: PropTypes.func,
	containerRef: PropTypes.any,
	viewAllLink: PropTypes.string,
	cardRender: PropTypes.func,
	isLoading: PropTypes.bool
}

export default Carousel
