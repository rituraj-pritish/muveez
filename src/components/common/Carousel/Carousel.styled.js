import styled from 'styled-components'
import theme from 'theme'

export const Title = styled.div`
  color: ${theme.colors.white};
  margin-bottom: ${theme.spacing()};
  font-size: ${theme.spacing(1.7)};
`

export const Wrapper = styled.div`
  /* height: 400px;

  .slick-slider {
    height: 100%;
  }

  .slick-list {
    height: 100%;
  }

  .slick-track {
    height: 100%;
  } */

  .slick-slide {
    display: flex;
    justify-content: center;
    /* width: 100%;
    height: 100%;

    & > div {
      width: 100%;
      height: 100%;
      & > div {
        width: 100%;
        height: 100%;
      }
    } */
  }

  // arrows
  .slick-arrow {
    z-index: 1;
    background: black;
    height: 100%;
    width: ${theme.spacing(2)};
    opacity: 0.3;
    border-top-right-radius: ${theme.borderRadius};
    border-bottom-right-radius: ${theme.borderRadius};
  }

  .slick-next {
    right: 0;
  }

  .slick-prev {
    left: 0;
  }
`
