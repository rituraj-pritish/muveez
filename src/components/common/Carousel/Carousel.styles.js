import styled from 'styled-components'
import theme from 'theme'

export const SliderWrapper = styled.div`
  position: relative;
  padding-bottom: 30%; // to maintain aspect ratio

  .slick-slider {
    height: 100%;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    max-width: ${theme.maxWidth};
    max-height: 360px;
  }

  .slick-list {
    height: 100%;
  }

  .slick-track {
    height: 100%;
  }

  .slick-slide {
    width: 100%;
    height: 100%;

    & > div {
      display: flex;
      justify-content: center;
      width: 100%;
      height: 100%;

      // card      
      & > div {
        width: 90%;
        height: 96%;
      }
    }
  }

  // arrows
  .slick-arrow {
    margin-top: -5px;
    z-index: 1;
    background: ${({ theme }) => theme.colors.textTertiary + '33'}; // opacity 33%
    height: 98%;
    width: ${theme.spacing(2)};
    display: flex !important;

    svg {
      padding: ${theme.spacing(0.25)};
    }

    &:before {
      content: '';
    }
  }

  .slick-next {
    right: 0;
    border-top-right-radius: ${theme.borderRadius};
    border-bottom-right-radius: ${theme.borderRadius};
  }

  .slick-prev {
    left: 0;
    border-top-left-radius: ${theme.borderRadius};
    border-bottom-left-radius: ${theme.borderRadius};
  }

  //media queries

  @media (min-width: 1200px) {
    padding-bottom: 30%;
  }

  @media (max-width: 1150px) {
    padding-bottom: 37%;
  }

  @media (max-width: 900px) {
    padding-bottom: 42%;

    // card      
    .slick-slide > div > div {
      width: 80%;
    }
  }

  @media (max-width: 750px) {
    padding-bottom: 45%;

    // card      
    .slick-slide > div > div {
      width: 85%;
    }
  }

  @media (max-width: 600px) {
    padding-bottom: 70%;

    // card      
    .slick-slide > div > div {
      width: 85%;
    }
  }

  @media (max-width: 400px) {
    padding-bottom: 85%;

    // card      
    .slick-slide > div > div {
      width: 87%;
    }
  }

  @media (hover:none) {
    .slick-arrow {
      visibility: hidden;
    }
  }
`

export const Wrapper = styled.div`
  position: relative;
`
