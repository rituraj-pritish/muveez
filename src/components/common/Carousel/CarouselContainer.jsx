import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'

import Carousel from './Carousel'
import usePositionOffset from 'hooks/usePositionOffset'
import useCarouselData from 'hooks/useCarouselData'

const CarouselContainer = ({ path, toggleOptions, ...rest }) => {
	const [ref, isTriggered] = usePositionOffset(300)
	const queries = useCarouselData(
		path ? [path] : toggleOptions.map(({ value }) => value)
	)
	const [selectedOptionIndex, setSelectedOptionIndex] = useState(0)
	const [data, setData] = useState(null)

	useEffect(() => {
		if (!isTriggered) return
		queries[selectedOptionIndex].refetch()
	}, [isTriggered])

	useEffect(() => {
		if (queries[selectedOptionIndex].isSuccess) {
			setData(queries[selectedOptionIndex].data.results)
		}
	}, [queries])

	const onToggleChange = val => {
		const idx = toggleOptions.findIndex(({ value }) => value === val)
		if(!queries[idx].isSuccess) {
			queries[idx].refetch()
		}
		setSelectedOptionIndex(idx)
	}

	return (
		<Carousel
			containerRef={ref}
			toggleOptions={toggleOptions}
			onToggleChange={onToggleChange}
			data={data}
			{...rest}
		/>
	)
}

CarouselContainer.propTypes = {
	path: PropTypes.string,
	toggleOptions: PropTypes.array
}

export default CarouselContainer
