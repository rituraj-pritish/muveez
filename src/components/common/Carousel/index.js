import Carousel from './CarouselContainer'

export default Carousel