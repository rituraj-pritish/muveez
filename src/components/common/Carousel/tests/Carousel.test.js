import userEvent from '@testing-library/user-event'
import React from 'react'
import { render, screen } from 'tests/utils'
import Carousel from '../index'
import data from './data'

describe('Carousel', () => {
	test('Renders carousel', async () => {
		render(
			<Carousel
				title='Movies'
				data={data}
			/>
		)

		expect(screen.getByText('Movies')).toBeInTheDocument()
		expect(screen.queryByText('View all')).not.toBeInTheDocument()
	})

	test('Renders empty message if no data', () => {
		const { container } = render(
			<Carousel
				title='Movies'
				data={[]}
			/>
		)

		expect(screen.getByText('Movies')).toBeInTheDocument()
		// react-slick renders two copies
		expect(screen.getAllByText('No data available')).toHaveLength(2)
		expect(screen.queryByText('View all')).not.toBeInTheDocument()

		expect(container).toMatchSnapshot()
	})

	test('Renders carousel with view all link', () => {
		const { history } = render(
			<Carousel
				title='Movies'
				data={data}
				viewAllLink='/link'
			/>
		)

		expect(screen.getByText('Movies')).toBeInTheDocument()
		expect(screen.getByText('View all')).toBeInTheDocument()

		jest.spyOn(history, 'push')
		userEvent.click(screen.getByText('View all'))
		expect(history.push).toHaveBeenCalledWith('/link')
	})
})
