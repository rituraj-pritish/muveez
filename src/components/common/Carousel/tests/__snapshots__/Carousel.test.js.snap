// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Renders empty message if no data 1`] = `
<div>
  <div
    class="css-ahmskn-SkeletonTheme"
  >
    <div
      class="sc-cidCJl bsHAMR"
    >
      <div />
      <div
        class="sc-dkPtyc cRDQcT"
      >
        <div>
          <div
            class="sc-hKwCoD gKFhOv"
            color="primary"
            font-size="4"
            font-weight="bold"
          >
            Movies
          </div>
        </div>
      </div>
      <div
        class="sc-cCcYRi cdLqPU"
      >
        <div
          class="slick-slider slick-initialized"
        >
          
          <div
            class="slick-list"
          >
            <div
              class="slick-track"
              style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
            >
              <div
                aria-hidden="false"
                class="slick-slide slick-active slick-current"
                data-index="0"
                style="outline: none; width: 0px;"
                tabindex="-1"
              >
                <div>
                  <div
                    style="width: 100%; display: inline-block;"
                    tabindex="-1"
                  >
                    <div
                      class="sc-dkPtyc dxsGev"
                      height="100%"
                    >
                      No data available
                    </div>
                  </div>
                </div>
              </div>
              <div
                aria-hidden="false"
                class="slick-slide slick-active slick-cloned"
                data-index="1"
                style="width: 0px;"
                tabindex="-1"
              >
                <div>
                  <div
                    style="width: 100%; display: inline-block;"
                    tabindex="-1"
                  >
                    <div
                      class="sc-dkPtyc dxsGev"
                      height="100%"
                    >
                      No data available
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          
          
        </div>
      </div>
    </div>
  </div>
</div>
`;
